File: Codility/Brackets.cpp
#include <iostream>
#include <algorithm>
#include <stack>
#include <string>
using namespace std;

int solution(string S) {
    stack<char> st;
    for (int i = 0; i < S.size(); i++) {
        if (st.empty())
            st.push(S[i]);
        else {
            if (S[i] == ')') {
                if (st.top() == '(')
                    st.pop();
                else
                    return 0;
            }
            else if (S[i] == ']') {
                if (st.top() == '[')
                    st.pop();
                else
                    return 0;
            }
            else if (S[i] == '}') {
                if (st.top() == '{')
                    st.pop();
                else
                    return 0;
            }
            else
                st.push(S[i]);
        }
    }
    if (st.empty()) return 1;
    else return 0;
}

int main()
{
    cout << solution("{}}{");
}

File: Java/src/codility/Brackets.java
package codility;

import java.util.Stack;

public class Brackets {
    class Solution {
        public int solution(String S) {
            Stack<Character> st = new Stack<>();
            for(int i=0;i<S.length();i++){
                if(S.charAt(i) == '(' || S.charAt(i) == '{' || S.charAt(i) == '[')
                    st.push(S.charAt(i));
                else if(st.empty()) return 0;
                else if(S.charAt(i) == ')' && st.peek()=='(') st.pop();
                else if(S.charAt(i) == '}' && st.peek()=='{') st.pop();
                else if(S.charAt(i) == ']' && st.peek()=='[') st.pop();
                else return 0;
            }
            if(!st.empty()) return 0;
            return 1;
        }
    }
}
